modules/setupadBidAdapter.js

```diff
@@ -0,0 +1,299 @@
+      queryParams.push('usp_consent=' + (uspConsent || ''));
+
+      const strQueryParams = queryParams.join('&');
+
+      syncs.push({
+        type: 'iframe',
+        url: SYNC_ENDPOINT + strQueryParams + '&type=iframe',
+      });
+
+      return syncs;
+    }
+
+    return [];
+  },
+
+  onBidWon: function (bid) {
+    let bidder = bid.bidder || bid.bidderCode;
+    const auctionId = bid.auctionId;
+    if (bidder !== BIDDER_CODE) return;
+
+    let params;
+    if (bid.params) {
+      params = Array.isArray(bid.params) ? bid.params : [bid.params];
+    } else {
+      if (Array.isArray(bid.bids)) {
+        params = bid.bids.map((singleBid) => singleBid.params);
+      }
+    }
+
+    if (!params?.length) return;
+
+    const placementIdsArray = [];
+    params.forEach((param) => {
+      if (!param.placement_id) return;
+      placementIdsArray.push(param.placement_id);
+    });
+
+    const placementIds = (placementIdsArray.length && placementIdsArray.join(';')) || '';
+
+    if (!placementIds) return;
+
+    let extraBidParams = '';
+
+    // Iterate through all bidders to find the winning bidder by using creativeId as identification
+    for (const bidderName in biddersCreativeIds) {
+      if (
+        biddersCreativeIds.hasOwnProperty(bidderName) &&
+        biddersCreativeIds[bidderName] === bid.creativeId
+      ) {
+        bidder = bidderName;
+        break; // Exit the loop if a match is found
+      }
+    }
+
+    // Add extra parameters
+    extraBidParams = `&cpm=${bid.originalCpm}&currency=${bid.originalCurrency}`;
+
+    const url = `${REPORT_ENDPOINT}?event=bidWon&bidder=${bidder}&placementIds=${placementIds}&auctionId=${auctionId}${extraBidParams}&timestamp=${Date.now()}`;
+    triggerPixel(url);
+  },
+};
+
+function getBidders(serverResponse) {
+  const bidders = serverResponse
+    .map((res) => Object.keys(res.body.ext.responsetimemillis || []))
+    .flat(1);
+
+  if (bidders.length) {
+    return encodeURIComponent(JSON.stringify([...new Set(bidders)]));
+  }
+}
+
+function getAd(bid) {
+  let ad, adUrl, vastXml, vastUrl;
+
+  switch (deepAccess(bid, 'ext.prebid.type')) {
+    default:
+      if (bid.adm && bid.nurl) {
+        ad = bid.adm;
+        ad += createTrackPixelHtml(decodeURIComponent(bid.nurl));
+      } else if (bid.adm) {
+        ad = bid.adm;
+      } else if (bid.nurl) {
+        adUrl = bid.nurl;
+      }
+  }
+
+  return { ad, adUrl, vastXml, vastUrl };
+}
+
+function getSiteObj() {
+  const refInfo = (getRefererInfo && getRefererInfo()) || {};
+
+  return {
+    page: refInfo.page,
+    ref: refInfo.ref,
+    domain: refInfo.domain,
+  };
+}
+
+function getDeviceObj() {
+  return {
+    w:
+      window.innerWidth ||
+      window.document.documentElement.clientWidth ||
+      window.document.body.clientWidth ||
+      0,
+    h:
+      window.innerHeight ||
+      window.document.documentElement.clientHeight ||
+      window.document.body.clientHeight ||
+      0,
+  };
```

Comment:
Both these device and site objects are probably available in the `ortb` object of your `bidderRequest`

Examples:

<img width="2" alt="image" src="https://github.com/prebid/Prebid.js/assets/13972794/b82f9c01-e3e7-49c9-b512-6e6aa1d649b0">


So maybe can simplify the logic to pick from there.